File: src/revenue.js
/* ---------- revenue ----------
   Itineraries, demand allocation and revenue, derived from the finished
   schedule. Like economics.js this only reads the model; nothing here feeds
   back into the engine, so the acceptance metrics and the ten checks cannot
   move because of it.

   Mirrors revenue.py. tools/rev_parity.js feeds both the same schedule and
   fails if they disagree.

   The point of the file: demand is origin-and-destination, cost is per leg,
   and in a bank structure a large share of passengers ride two legs on one
   fare. Multiplying per-leg loads by per-leg fares would invent revenue that
   was never sold. So this builds the itineraries the schedule offers,
   allocates each market's demand across them, and prorates the fare back onto
   the legs that carried it. Leg revenue sums to market revenue by
   construction. */

const INTL = (()=>{ try { return JSON.parse(document.getElementById("intl").textContent); }
                    catch(e){ return {demandFactor:1, fareCurve:{}, overrides:{}}; } })();
const INTL_FACTOR = INTL.demandFactor != null ? INTL.demandFactor : 1;
const INTL_FARE = INTL.fareCurve || {};
const INTL_OVERRIDE = INTL.overrides || {};

const REV_MAX_CONNECT = 240;   // minutes; beyond this nobody buys it
const REV_CIRCUITY = 1.35;     // connecting path vs nonstop great circle
const REV_CONNECT_PENALTY = 0.65;
const REV_COMPETITION = 1.0;   // QSI of every other airline in a market
const REV_SPILL_PASSES = 4;
/* What a seat is worth against economy, by cabin AND by the kind of seat in it.
   A domestic recliner up front and a lie-flat suite are not the same product and
   must not earn the same fare: the seatmap can tell them apart, so the revenue
   model has to as well, or a 20-suite A321XLR earns exactly what a 20-recliner
   A321 earns. */
const SEAT_MULT = {
  F:  {lieflat: 5.5, recliner: 3.0},
  PE: {recliner: 2.0, standard: 1.6},
  Y:  {standard: 1.0}
};

/* A flat bed is worth paying for because you can sleep on it, so its premium
   depends on how long the flight is. Below the floor it is a wide seat you
   cannot use as a bed; by the reference stage it earns the full multiplier.
   Without this, suites on a 400 nm hop would pay as well as suites on a
   transcon, which is the kind of answer that discredits a model. Recliners and
   extra-legroom seats are not tapered: their value is legroom, and you get that
   on any stage. */
const LIEFLAT_FULL_NM = 1800, LIEFLAT_FLOOR = 0.30;

function seatMult(cabin, kind, nm){
  const m = (SEAT_MULT[cabin] || {})[kind];
  if(m === undefined) return 1;
  if(kind !== "lieflat") return m;
  const t = Math.min(1, Math.max(LIEFLAT_FLOOR, (nm || 0) / LIEFLAT_FULL_NM));
  return 1 + (m - 1) * t;
}

/* The raw, un-normalised worth of an average seat on one flight. */
function rawPremium(f){
  const s = SPEC[f.t]; if(!s) return 1;
  const tot = (s.F||0) + (s.PE||0) + (s.Y||0); if(!tot) return 1;
  let v = 0;
  for(const c of ["F","PE","Y"])
    v += seatMult(c, seatKindOf(s.geom, c), f.nm) * (s[c]||0);
  return v / tot;
}

const isIntl = (o,d) => (AP[o] && AP[o][5]) !== "United States"
                     || (AP[d] && AP[d][5]) !== "United States";

/* Fare against distance, fitted to the real DB1C fares. Used where a market
   has passengers but no fare of its own. Weak — r2 about 0.19 — but better
   than a flat number. */
const FARE_FIT = (()=>{
  if(!DOT || !DOT.rows) return {a:100, b:0, r2:0, n:0};
  let n=0, sx=0, sy=0;
  const pts=[];
  for(const k in DOT.rows){
    const v=DOT.rows[k], o=k.slice(0,3), d=k.slice(3);
    if(!AP[o] || !AP[d] || !v[1]) continue;
    const nm=dist(o,d); if(nm<50) continue;
    const x=Math.log(nm), y=Math.log(v[1]);
    pts.push([x,y]); sx+=x; sy+=y; n++;
  }
  if(!n) return {a:100, b:0, r2:0, n:0};
  const mx=sx/n, my=sy/n;
  let sxy=0, sxx=0;
  for(const [x,y] of pts){ sxy+=(x-mx)*(y-my); sxx+=(x-mx)*(x-mx); }
  const b=sxy/sxx, a=my-b*mx;
  let ss=0, rs=0;
  for(const [x,y] of pts){ ss+=(y-my)*(y-my); rs+=(y-(a+b*x))*(y-(a+b*x)); }
  return {a:Math.exp(a), b, r2:1-rs/ss, n};
})();

/* Passengers per direction per day, average fare, and where they came from:
   "db1c" measured, "estimated" international, "subthreshold" a US market too
   small for the DB1C sample, "none" unsizeable. Provenance travels with the
   number so an estimate is never reported as a measurement. */
function marketOf(o,d){
  const k = o<d ? o+d : d+o;
  const ov = INTL_OVERRIDE[k];
  if(ov) return {pax:ov[0], fare:ov[1], src:"db1c"};
  const r = DOT && DOT.rows ? DOT.rows[k] : null;
  const nm = dist(o,d);
  if(r) return {pax:r[0], fare:r[1] || FARE_FIT.a*Math.pow(nm,FARE_FIT.b), src:"db1c"};
  const so=DEM.size[o], sd=DEM.size[d];
  if(!so || !sd || nm<50) return {pax:0, fare:0, src:"none"};
  const air = 1/(1+Math.exp(-(nm-190)/55));
  const grav = so*sd/Math.pow(nm,DEM.beta)*DEM.k*air;
  // The Caribbean calibration applies to international markets. A US market
  // absent from DB1C is not international, it is under the sample threshold,
  // and multiplying it would invent traffic.
  if(isIntl(o,d))
    return {pax: grav*INTL_FACTOR, src:"estimated",
            fare: (INTL_FARE.a || FARE_FIT.a)*Math.pow(nm, INTL_FARE.b != null ? INTL_FARE.b : FARE_FIT.b)};
  return {pax:grav, fare:FARE_FIT.a*Math.pow(nm,FARE_FIT.b), src:"subthreshold"};
}

/* Nonstops, plus one-stop connections over the stations. Times are UTC
   minutes on a repeating day, so a bank wrapping past midnight still
   connects — the +1440 shift is tried as well as the same-day window. */
function buildItineraries(F){
  const itins = new Map(), push = (o,d,it) => {
    const k=o+"|"+d; let a=itins.get(k); if(!a){ a=[]; itins.set(k,a); } a.push(it); };
  const arrAt = new Map(), depAt = new Map();
  for(const f of F){
    push(f.o, f.d, {legs:[f], nm:f.nm, elapsed:f.blkMin, stops:0});
    if(STA.includes(f.d)){ if(!arrAt.has(f.d)) arrAt.set(f.d,[]); arrAt.get(f.d).push(f); }
    if(STA.includes(f.o)){ if(!depAt.has(f.o)) depAt.set(f.o,[]); depAt.get(f.o).push(f); }
  }
  for(const s of STA){
    const A=arrAt.get(s)||[], B=depAt.get(s)||[];
    for(const a of A) for(const b of B){
      if(b.d === a.o) continue;                  // a return, not a connection
      const aArr = (a.depX != null ? a.depX : a.depU) + a.blkMin;
      let wait = null;
      for(const sh of [0,1440]){
        const w = (b.depX != null ? b.depX : b.depU) + sh - aArr;
        if(w >= MCT && w <= REV_MAX_CONNECT){ wait = w; break; }
      }
      if(wait === null) continue;
      const nm = a.nm + b.nm, gc = dist(a.o, b.d);
      if(gc < 50 || nm > REV_CIRCUITY*gc) continue;
      push(a.o, b.d, {legs:[a,b], nm, stops:1,
                      elapsed: a.blkMin + wait + b.blkMin});
    }
  }
  return itins;
}

/* Split each market's demand across its itineraries, then spill.
   `competition` is the QSI attributed to every other airline in a market, so
   our share is Q/(Q+competition): a market we serve with three nonstops takes
   more than one we only reach over a hub. */
/* Premiums normalised so the seat-weighted fleet mean is 1.0, which keeps the
   network total pinned to the measured DB1C fares — cabin mix moves revenue
   between flights rather than adding any.

   Per FLIGHT rather than per type, because the lie-flat taper makes the same
   aircraft worth more on a long sector than a short one.

   ASSUMPTION, unchanged and still the weak point: passengers are spread across
   cabins in proportion to seats, so a premium cabin is credited whether or not
   anyone would have bought it. Cabin-level load factors and a cabin-aware spill
   would fix it. Mirrors flight_premiums() in revenue.py. */
function flightPremiums(F){
  let num = 0, den = 0;
  const raw = new Map();
  for(const f of F){
    const s = SPEC[f.t]; if(!s) continue;
    const seats = (s.F||0)+(s.PE||0)+(s.Y||0);
    const r = rawPremium(f);
    raw.set(f.id, r); num += r*seats; den += seats;
  }
  const mean = den ? num/den : 1;
  const prem = new Map();
  for(const [id, r] of raw) prem.set(id, mean ? r/mean : 1);
  return {prem, mean};
}

function allocateDemand(F, itins, competition, prem){
  const boarded=new Map(), legrev=new Map(), estrev=new Map(), perMarket=[];
  const st={pax:0, conn:0, rev:0, spilled:0, markets:0, real:0,
            estMarkets:0, estPax:0, estRev:0};
  for(const [key, opts] of itins){
    const [o,d] = key.split("|");
    const m = marketOf(o,d);
    if(m.pax <= 0) continue;
    let best = Infinity;
    for(const i of opts) if(i.elapsed < best) best = i.elapsed;
    let tq = 0;
    for(const i of opts){
      i.qsi = Math.pow(best/i.elapsed, 2) * (i.stops ? REV_CONNECT_PENALTY : 1);
      tq += i.qsi;
    }
    const share = tq/(tq+competition);
    for(const i of opts) i.pax = m.pax*share*i.qsi/tq;
    st.markets++;
    if(m.src === "db1c") st.real++;
    if(m.src === "estimated") st.estMarkets++;
    perMarket.push({o, d, fare:m.fare, src:m.src, opts});
  }
  // Spill. Capping one flight frees demand on another, so a few passes.
  const keep = new Map(); for(const f of F) keep.set(f.id, 1);
  for(let p=0; p<REV_SPILL_PASSES; p++){
    boarded.clear();
    for(const m of perMarket) for(const i of m.opts){
      let k = 1; for(const f of i.legs) k = Math.min(k, keep.get(f.id));
      for(const f of i.legs) boarded.set(f.id, (boarded.get(f.id)||0) + i.pax*k);
    }
    for(const f of F){
      const b = boarded.get(f.id)||0;
      if(b > f.seats) keep.set(f.id, keep.get(f.id)*f.seats/b);
    }
  }
  boarded.clear();
  for(const m of perMarket) for(const i of m.opts){
    let k = 1; for(const f of i.legs) k = Math.min(k, keep.get(f.id));
    const flown = i.pax*k;
    // Kept on the itinerary so the planning surfaces can ask which connections
    // the schedule actually realised, rather than only how many there were.
    // Numerically inert: nothing downstream reads it.
    i.flown = flown;
    st.spilled += i.pax - flown;
    st.pax += flown;
    if(i.stops) st.conn += flown;
    // Distance-weighted mean of the legs' cabin premiums. Weighting by distance
    // and not by leg count is deliberate: the fare is prorated onto the legs by
    // distance a few lines below, so using the same weight here keeps leg
    // revenue summing to itinerary revenue exactly — the invariant this whole
    // file exists to hold.
    let im = 0;
    for(const f of i.legs) im += (prem.get(f.id) || 1)*f.nm;
    im /= i.nm;
    const rev = flown*m.fare*im;
    st.rev += rev;
    if(m.src === "estimated"){ st.estPax += flown; st.estRev += rev; }
    for(const f of i.legs){
      boarded.set(f.id, (boarded.get(f.id)||0) + flown);
      const share = rev*f.nm/i.nm;                 // straight distance proration
      legrev.set(f.id, (legrev.get(f.id)||0) + share);
      if(m.src === "estimated") estrev.set(f.id, (estrev.get(f.id)||0) + share);
    }
  }
  return {boarded, legrev, estrev, stats:st};
}

function revenueModel(M){
  const F = M.flights.map(f=>Object.assign({}, f,
    {seats: (SPEC[f.t] && SPEC[f.t].seats) || 0}));
  const itins = buildItineraries(F);
  const cabins = flightPremiums(F);
  const prem = cabins.prem;
  const r = allocateDemand(F, itins, REV_COMPETITION, prem);
  const seats = F.reduce((s,f)=>s+f.seats, 0);
  // The ceiling: what this schedule fills taking every passenger in every
  // market it touches. Below 100% is capacity that cannot fill unopposed.
  const ceiling = seats
    ? [...allocateDemand(F, buildItineraries(F), 0, prem).boarded.values()]
        .reduce((a,b)=>a+b,0)/seats : 0;
  const asm = F.reduce((s,f)=>s+f.seats*f.nm, 0)*SM;
  const rpm = F.reduce((s,f)=>s+(r.boarded.get(f.id)||0)*f.nm, 0)*SM;
  let nonstop=0, onestop=0;
  for(const opts of itins.values()) for(const i of opts) i.stops ? onestop++ : nonstop++;
  return Object.assign(r, {
    itins,                       // for connectivity analysis; see ui/connectivity.js
    itineraries:{nonstop, onestop}, seats, asm, rpm, ceiling,
    lf: seats ? [...r.boarded.values()].reduce((a,b)=>a+b,0)/seats : 0,
    rasm: asm ? r.stats.rev/asm*100 : 0,
    yield: rpm ? r.stats.rev/rpm*100 : 0,
    cabins: {mean: cabins.mean}, fareFit: FARE_FIT
  });
}

File: revenue.py
#!/usr/bin/env python3
"""Revenue, with connections. Reference implementation; src/revenue.js mirrors it.

    python3 make_legs.py && python3 engine.py && python3 revenue.py

WHY THIS EXISTS. Demand is origin-and-destination, cost is per leg. In a bank
structure a large share of passengers ride two legs on one fare, so multiplying
per-leg loads by per-leg fares would invent revenue that was never sold. This
builds the itineraries the schedule actually offers, allocates each market's
demand across them, and then prorates each fare back onto the legs that carried
it. Leg revenue therefore sums to market revenue by construction.

FOUR MODELLING CHOICES, all of them assumptions rather than measurements:

  1. CONNECTION LEGALITY. A connection is a bank arrival followed by a
     departure at the same station, at least MCT apart and no more than
     MAX_CONNECT, with the two legs no more circuitous than CIRCUITY times
     the nonstop great circle. Only stations connect; a spoke is not a hub.

  2. ALLOCATION. A quality-of-service index. Each itinerary scores on elapsed
     time against the market's best, with a flat penalty for connecting;
     market demand splits in proportion to score. This is the standard shape,
     and it is calibrated against nothing.

  3. PRORATION. Straight distance proration: a leg earns the share of the fare
     matching its share of the itinerary's miles. Real airlines use weighted
     schemes that favour the local leg; distance is the transparent default.

  4. MARKET SHARE. DB1C reports the WHOLE market, every carrier in it. The
     first version of this file allocated all of it to us and spilled eight
     times more traffic than it carried, which is what a one-airline-owns-
     everything assumption looks like when it hits a capacity constraint.
     Share is now QSI against an unobserved competitor, share =
     Q / (Q + COMPETITION), so markets where the schedule is strong win more.
     COMPETITION is not measurable from anything here, so it is a single named
     parameter: 1.0 means a rival of equal schedule quality, and we split the
     market evenly. Load factor is then an OUTPUT, not a target. An earlier
     draft solved COMPETITION to hit an 83% load factor and it drove to zero,
     because even taking 100% of every market this schedule only fills 63.5%.
     That is a finding about the network, not a number to calibrate away:
     capacity sits in thin markets that cannot fill at any plausible share.

  5. SPILL. A flight cannot board more than it has seats. Demand over capacity
     is spilled with no recapture onto other flights, so this understates
     revenue on tight markets rather than flattering it. Applied over a few
     passes, since capping one flight frees demand on another.

DATA PROVENANCE. Every market is tagged. `db1c` is measured US O&D. `estimated`
is international, where gravity demand is scaled by a factor calibrated on the
DB1C-domestic Caribbean (see tools/intl_calibrate.py) and fares come from a
curve fitted on the same markets. Totals are reported both ways, because an
estimated route showing a loss is a statement about the estimate, not about
the route. Real international data backfills through the `overrides` table in
src/data/intl_demand.json without touching this file.

CABIN FARES reconcile down to the observed DB1C average rather than building
up from it: multipliers are normalised so the seat-weighted mean equals 1.0,
which means the real fare data constrains the answer instead of decorating it.
"""
import json, math, pathlib
from collections import defaultdict

ROOT = pathlib.Path(__file__).resolve().parent
load = lambda p: json.loads((ROOT / p).read_text())

AP = load("src/data/airports.json")
DEM = load("src/data/dot_db1c.json")
DOT = DEM.get("dot") or {}
ROWS = DOT.get("rows", {})
SIZE = DEM.get("size", {})
BETA, KGRAV = DEM.get("beta", 1.2), DEM.get("k", 1.0)
FLEETDATA = load("src/data/fleet.json")
FLEET = {f["t"]: f for f in FLEETDATA["types"]}
CFG = load("src/data/config.json")
INTL = load("src/data/intl_demand.json")
INTL_OVERRIDE = INTL.get("overrides", {})
INTL_FACTOR = INTL.get("demandFactor", 1.0)
INTL_FARE = INTL.get("fareCurve", {})
STA = set(load("src/data/stations.json"))
LINES = load("lines.json")["lines"]

MCT = CFG.get("mct", 40)
MAX_CONNECT = 240        # minutes; beyond this nobody buys the itinerary
CIRCUITY = 1.35          # connecting path vs nonstop great circle
CONNECT_PENALTY = 0.65   # QSI multiplier for a one-stop against a nonstop
COMPETITION = 1.0        # QSI attributed to every other airline in a market
SPILL_PASSES = 4
# Cabin fare multipliers, normalised below so the seat-weighted mean is 1.0.
CABIN_MULT = {"F": 3.0, "PE": 1.6, "Y": 1.0}

US = "United States"
is_intl = lambda o, d: (AP.get(o, [None]*6)[5] != US
                        or AP.get(d, [None]*6)[5] != US)

DIST = {}


def dist(a, b):
    k = (a, b) if a < b else (b, a)
    if k not in DIST:
        la1, lo1 = math.radians(AP[a][2]), math.radians(AP[a][3])
        la2, lo2 = math.radians(AP[b][2]), math.radians(AP[b][3])
        DIST[k] = 3440.065 * 2 * math.asin(math.sqrt(
            math.sin((la2 - la1) / 2) ** 2 + math.cos(la1) * math.cos(la2)
            * math.sin((lo2 - lo1) / 2) ** 2))
    return DIST[k]


def fare_curve():
    """Fare against distance, fitted to the real DB1C fares.

    Markets outside the DB1C sample -- everything international -- have
    passengers from the gravity model but no fare at all. Rather than leave
    them at zero or invent a flat number, fit log(fare) on log(nm) across the
    12,241 real markets and use the curve where a real fare is missing.
    """
    xs, ys = [], []
    for k, v in ROWS.items():
        o, d = k[:3], k[3:]
        if o not in AP or d not in AP or not v[1]:
            continue
        nm = dist(o, d)
        if nm < 50:
            continue
        xs.append(math.log(nm))
        ys.append(math.log(v[1]))
    n = len(xs)
    mx, my = sum(xs) / n, sum(ys) / n
    b = (sum((x - mx) * (y - my) for x, y in zip(xs, ys))
         / sum((x - mx) ** 2 for x in xs))
    a = my - b * mx
    ss = sum((y - my) ** 2 for y in ys)
    rs = sum((y - (a + b * x)) ** 2 for x, y in zip(xs, ys))
    return math.exp(a), b, 1 - rs / ss, n


FARE_A, FARE_B, FARE_R2, FARE_N = fare_curve()


def market(o, d):
    """Passengers per direction per day, average fare, and where they came from.

    Returns provenance as well as numbers: "db1c" measured, "estimated" for
    international, "none" where the market cannot be sized at all.
    """
    k = o + d if o < d else d + o
    ov = INTL_OVERRIDE.get(k)
    if ov:                                   # real data, backfilled
        return ov[0], ov[1], "db1c"
    r = ROWS.get(k)
    if r:
        return r[0], (r[1] or FARE_A * dist(o, d) ** FARE_B), "db1c"
    so, sd = SIZE.get(o), SIZE.get(d)
    nm = dist(o, d)
    if not so or not sd or nm < 50:
        return 0.0, 0.0, "none"
    air = 1 / (1 + math.exp(-(nm - 190) / 55))
    grav = so * sd / nm ** BETA * KGRAV * air
    # The Caribbean calibration applies to international markets. A US market
    # absent from DB1C is not international, it is just under the sample
    # threshold of roughly three passengers a day, and multiplying it by 7.56
    # would invent traffic.
    if is_intl(o, d):
        return (grav * INTL_FACTOR,
                INTL_FARE.get("a", FARE_A) * nm ** INTL_FARE.get("b", FARE_B),
                "estimated")
    return grav, FARE_A * nm ** FARE_B, "subthreshold"


def flights():
    """Every departure, in UTC minutes, with seats and cabin split."""
    out = []
    for li, l in enumerate(LINES):
        for gi, (o, d, dep, arr, b) in enumerate(l["flights"]):
            f = FLEET[l["type"]]
            out.append(dict(id=li * 100 + gi, o=o, d=d, t=l["type"],
                            dep=dep, arr=arr, blk=b / 60.0, nm=dist(o, d),
                            seats=f["F"] + f["PE"] + f["Y"],
                            cabins={c: f[c] for c in ("F", "PE", "Y")}))
    return out


def itineraries(F):
    """Nonstops, plus one-stop connections over the stations.

    Times are UTC minutes on a repeating day, so a bank that wraps past
    midnight still connects -- the +1440 shift is tried as well as the
    same-day window.
    """
    itins = defaultdict(list)
    for f in F:
        itins[(f["o"], f["d"])].append(
            {"legs": [f], "nm": f["nm"], "elapsed": f["arr"] - f["dep"],
             "stops": 0})
    arr_at, dep_at = defaultdict(list), defaultdict(list)
    for f in F:
        if f["d"] in STA:
            arr_at[f["d"]].append(f)
        if f["o"] in STA:
            dep_at[f["o"]].append(f)
    for s in STA:
        for a in arr_at[s]:
            for b in dep_at[s]:
                if b["d"] == a["o"]:
                    continue          # not a connection, a return
                for shift in (0, 1440):
                    wait = b["dep"] + shift - a["arr"]
                    if MCT <= wait <= MAX_CONNECT:
                        break
                else:
                    continue
                nm = a["nm"] + b["nm"]
                gc = dist(a["o"], b["d"])
                if gc < 50 or nm > CIRCUITY * gc:
                    continue
                itins[(a["o"], b["d"])].append(
                    {"legs": [a, b], "nm": nm, "stops": 1,
                     "elapsed": a["arr"] - a["dep"] + wait
                                + b["arr"] - b["dep"]})
    return itins


def cabin_shares():
    """Fare multipliers normalised so the fleet-weighted mean is 1.0."""
    seats = defaultdict(float)
    for l in LINES:
        f = FLEET[l["type"]]
        for c in ("F", "PE", "Y"):
            seats[c] += f[c] * len(l["flights"])
    tot = sum(seats.values())
    mean = sum(CABIN_MULT[c] * seats[c] / tot for c in seats)
    return {c: CABIN_MULT[c] / mean for c in seats}, \
           {c: seats[c] / tot for c in seats}


def allocate(itins, competition):
    """Split each market's demand across its itineraries, then spill.

    `competition` is the QSI attributed to every other airline in a market.
    Our share is Q/(Q+competition), so a market where we fly three nonstops
    takes a larger share than one we only reach over a hub.
    """
    boarded = defaultdict(float)          # flight id -> passengers
    legrev = defaultdict(float)           # flight id -> prorated revenue
    estrev = defaultdict(float)           # of which from estimated markets
    stats = dict(pax=0.0, conn=0.0, rev=0.0, spilled=0.0,
                 markets=0, real=0, nonstopOnly=0,
                 estPax=0.0, estRev=0.0, estMarkets=0)
    src_of = {}
    per_market = []
    for (o, d), opts in itins.items():
        pax, fare, src = market(o, d)
        if pax <= 0:
            continue
        src_of[(o, d)] = src
        best = min(i["elapsed"] for i in opts)
        for i in opts:
            i["qsi"] = ((best / i["elapsed"]) ** 2
                        * (CONNECT_PENALTY if i["stops"] else 1.0))
        tq = sum(i["qsi"] for i in opts)
        stats["markets"] += 1
        stats["real"] += 1 if src == "db1c" else 0
        stats["estMarkets"] += 1 if src == "estimated" else 0
        if all(i["stops"] == 0 for i in opts):
            stats["nonstopOnly"] += 1
        share = tq / (tq + competition)
        for i in opts:
            i["pax"] = pax * share * i["qsi"] / tq
        per_market.append(((o, d), pax, fare, opts, src))

    # Spill. Capping one flight frees demand on another, so a few passes.
    keep = {f["id"]: 1.0 for f in FLIGHTS}
    for _ in range(SPILL_PASSES):
        boarded.clear()
        for _, _, _, opts, _ in per_market:
            for i in opts:
                k = min(keep[f["id"]] for f in i["legs"])
                for f in i["legs"]:
                    boarded[f["id"]] += i["pax"] * k
        for f in FLIGHTS:
            b = boarded[f["id"]]
            if b > f["seats"]:
                keep[f["id"]] *= f["seats"] / b

    boarded.clear()
    for (o, d), pax, fare, opts, src in per_market:
        for i in opts:
            k = min(keep[f["id"]] for f in i["legs"])
            flown = i["pax"] * k
            stats["spilled"] += i["pax"] - flown
            stats["pax"] += flown
            if i["stops"]:
                stats["conn"] += flown
            rev = flown * fare
            stats["rev"] += rev
            if src == "estimated":
                stats["estPax"] += flown
                stats["estRev"] += rev
            for f in i["legs"]:
                boarded[f["id"]] += flown
                # straight distance proration
                legrev[f["id"]] += rev * f["nm"] / i["nm"]
                if src == "estimated":
                    estrev[f["id"]] += rev * f["nm"] / i["nm"]
    return boarded, legrev, stats, estrev


FLIGHTS = flights()


def main():
    F = FLIGHTS
    itins = itineraries(F)
    mult, seatshare = cabin_shares()
    seats = sum(f["seats"] for f in F)
    boarded, legrev, st, estrev = allocate(itins, COMPETITION)

    # The ceiling: what this schedule would fill if it took every passenger in
    # every market it touches. Anything below 100% is capacity the network
    # cannot fill even unopposed.
    ceiling = sum(allocate(itins, 0.0)[0].values()) / seats

    asm = sum(f["seats"] * f["nm"] for f in F) * 1.15078
    rpm = sum(boarded[f["id"]] * f["nm"] for f in F) * 1.15078
    nonstop = sum(len([i for i in v if not i["stops"]]) for v in itins.values())
    onestop = sum(len([i for i in v if i["stops"]]) for v in itins.values())

    print(f"fare curve: fare = {FARE_A:.2f} x nm^{FARE_B:.3f}  "
          f"(r2 {FARE_R2:.2f} on {FARE_N:,} real markets)")
    print(f"cabin multipliers, normalised: "
          + ", ".join(f"{c} {mult[c]:.2f}" for c in ("F", "PE", "Y"))
          + "  seat share "
          + ", ".join(f"{c} {seatshare[c]*100:.0f}%" for c in ("F", "PE", "Y")))
    print(f"competition QSI {COMPETITION:.2f} "
          f"(1.0 = one rival of equal schedule quality)")
    print(f"\nitineraries: {nonstop:,} nonstop, {onestop:,} one-stop over "
          f"{len(STA)} stations")
    print(f"markets served: {st['markets']:,} "
          f"({st['real']:,} measured DB1C, {st['estMarkets']:,} estimated "
          f"international at x{INTL_FACTOR})")
    print(f"\npassengers/day   {st['pax']:12,.0f}")
    print(f"  connecting     {st['conn']:12,.0f}  "
          f"({st['conn']/st['pax']*100:.0f}% of passengers)")
    print(f"  spilled        {st['spilled']:12,.0f}")
    print(f"revenue/day      {st['rev']:12,.0f}")
    print(f"  measured       {st['rev']-st['estRev']:12,.0f}  "
          f"({(1-st['estRev']/st['rev'])*100:.0f}%)")
    print(f"  estimated      {st['estRev']:12,.0f}  "
          f"({st['estRev']/st['rev']*100:.0f}%, international)")
    print(f"seats/day        {seats:12,.0f}")
    print(f"load factor      {sum(boarded.values())/seats*100:12.1f}%")
    print(f"  unopposed ceiling{ceiling*100:11.1f}%  "
          f"<- at 100% share of every market it serves")
    print(f"RASM (cents)     {st['rev']/asm*100:12.2f}")
    print(f"yield  (cents)   {st['rev']/rpm*100:12.2f}")
    byroute = defaultdict(lambda: [0.0, 0.0, 0.0])   # pax, seats, revenue
    for f in F:
        k = f["o"] + "-" + f["d"] if f["o"] < f["d"] else f["d"] + "-" + f["o"]
        a = byroute[k]
        a[0] += boarded[f["id"]]; a[1] += f["seats"]; a[2] += legrev[f["id"]]
    lfs = sorted(((v[0] / v[1], k, v) for k, v in byroute.items()
                  if v[1] > 0))
    print(f"\nthinnest 8 routes by load factor")
    for lf, k, v in lfs[:8]:
        print(f"  {k:9}{lf*100:6.1f}%{v[1]:8,.0f} seats{v[2]:11,.0f} rev/day")
    print(f"fullest 4")
    for lf, k, v in lfs[-4:]:
        print(f"  {k:9}{lf*100:6.1f}%{v[1]:8,.0f} seats{v[2]:11,.0f} rev/day")

    summary = {"pax": st["pax"], "rev": st["rev"], "conn": st["conn"],
               "spilled": st["spilled"], "estRev": st["estRev"],
               "estPax": st["estPax"], "markets": st["markets"],
               "real": st["real"], "estMarkets": st["estMarkets"],
               "seats": seats, "asm": asm, "rpm": rpm, "ceiling": ceiling,
               "nonstop": nonstop, "onestop": onestop,
               "lf": sum(boarded.values())/seats}
    (ROOT / "rev_py.json").write_text(json.dumps(summary, indent=1))
    json.dump({"boarded": {str(k): v for k, v in boarded.items()},
               "legrev": {str(k): v for k, v in legrev.items()},
               "estrev": {str(k): v for k, v in estrev.items()},
               "stats": st, "competition": COMPETITION, "ceiling": ceiling,
               "fareCurve": [FARE_A, FARE_B, FARE_R2]},
              open(ROOT / "revenue.json", "w"))
    print("\nwrote revenue.json")


if __name__ == "__main__":
    main()
